reclaimer/meta/wrappers/src/byteswapping_ext.c

```diff
@@ -0,0 +1,144 @@
+            swap[off + 6] = orig[off + 1];
+            swap[off + 7] = orig[off];
+        }
+    }
+}
+
+
+static void byteswap_four_byte_array(
+    char *orig, char *swap, uint64 start, uint64 end, uint64 struct_size,
+    int *offs, int off_count)
+{
+    for (int i = 0; i < off_count; i++) {
+        for (int off = start + offs[i]; off + 4 <= end; off += struct_size) {
+            swap[off]     = orig[off + 3];
+            swap[off + 1] = orig[off + 2];
+            swap[off + 2] = orig[off + 1];
+            swap[off + 3] = orig[off];
+        }
+    }
+}
+
+
+static void byteswap_two_byte_array(
+    char *orig, char *swap, uint64 start, uint64 end, uint64 struct_size,
+    int *offs, int off_count)
+{
+    for (int i = 0; i < off_count; i++) {
+        for (int off = start + offs[i]; off + 2 <= end; off += struct_size) {
+            swap[off]     = orig[off + 1];
+            swap[off + 1] = orig[off];
+        }
+    }
+}
+
+
+static PyObject *py_byteswap_struct_array(PyObject *self, PyObject *args) {
+    Py_buffer bufs[2];
+    Py_buffer field_offset_bufs[3];
+    int field_offset_sizes[3] = { 2, 4, 8 };
+    uint64 start, end, struct_size;
+
+    if (!PyArg_ParseTuple(args, "y*w*KKK|y*y*y*:byteswap_struct_array",
+        &bufs[0], &bufs[1], &start, &end, &struct_size,
+        &field_offset_bufs[0], &field_offset_bufs[1], &field_offset_bufs[2])) {
+        return Py_BuildValue("");  // return Py_None while incrementing it
+    }
+
+    // handle invalid data sizes and such
+    if (field_offset_bufs[0].itemsize != 4) {
+        PySys_FormatStdout("two_byte_offs array format must be 'I'.\n");
+    } else if (field_offset_bufs[1].itemsize != 4) {
+        PySys_FormatStdout("four_byte_offs array format must be 'I'.\n");
+    } else if (field_offset_bufs[2].itemsize != 4) {
+        PySys_FormatStdout("eight_byte_offs array format must be 'I'.\n");
+    } else if (end > bufs[0].len) {
+        PySys_FormatStdout("end must within bounds of input buffer.\n");
+    } else if (bufs[0].buf == bufs[1].buf) {
+        PySys_FormatStdout("src and dst cannot be the same.\n");
+    } else if (field_offset_bufs[2].itemsize != 4) {
+        PySys_FormatStdout("eight_byte_offs array format must be 'I'.\n");
+    } else if (struct_size == 0) {
+        PySys_FormatStdout("struct size must be greater than zero.\n");
+    } else if (start > end) {
+        PySys_FormatStdout("start must be less than end.\n");
+    } else {
+        int item_count;
+        for (int i = 0; i < 3; i++) {
+            if (field_offset_bufs[i].buf == NULL)
+                continue;
+
+            item_count = field_offset_bufs[i].len / field_offset_bufs[i].itemsize;
+            for (int j = 0; j < item_count; j++)
+                if (((int *)field_offset_bufs[i].buf)[j] + field_offset_sizes[i] > struct_size) {
+                    PySys_FormatStdout("offets must be within the struct size.\n");
+                    goto exit;
+                }
```

Comment:
Inconsistent indentation style. Looks confusing.

Reply:
You're right. gonna put another set of curly braces